pkg/svc/svc.go

```diff
@@ -0,0 +1,486 @@
+
+		if !k8sSVCFrontends.LooseMatch(svc.FE.L3n4Addr) {
+			log.WithFields(logrus.Fields{
+				logfields.ServiceID: svc.FE.ID,
+				logfields.L3n4Addr:  logfields.Repr(svc.FE.L3n4Addr)}).
+				Warn("Deleting no longer present service")
+
+			if err := s.deleteServiceLocked(svc); err != nil {
+				return fmt.Errorf("Unable to remove service %+v: %s", svc, err)
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func (s *Service) restoreBackendsLocked() error {
+	backends, err := lbmap.DumpBackendMapsToUserspace()
+	if err != nil {
+		return fmt.Errorf("Unable to dump backend maps: %s", err)
+	}
+
+	for _, b := range backends {
+		if err := service.RestoreBackendID(b.L3n4Addr, b.ID); err != nil {
+			return fmt.Errorf("Unable to restore backend ID %d for %q: %s",
+				b.ID, b.L3n4Addr, err)
+		}
+
+		hash := b.L3n4Addr.SHA256Sum()
+		s.backendByHash[hash] = b
+	}
+
+	return nil
+}
+
+func (s *Service) deleteOrphanBackends() error {
+	for hash, b := range s.backendByHash {
+		if s.backendRefCount[hash] == 0 {
+			service.DeleteBackendID(b.ID)
+			if err := lbmap.DeleteBackendByID(uint16(b.ID), b.L3n4Addr.IsIPv6()); err != nil {
+				return fmt.Errorf("Unable to remove backend %d from map: %s", b.ID, err)
+			}
+			delete(s.backendByHash, hash)
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) restoreServicesLocked() error {
+	failed, restored := 0, 0
+
+	_, svcs, errors := lbmap.DumpServiceMapsToUserspaceV2()
+	for _, err := range errors {
+		log.WithError(err).Warning("Error occurred while dumping service maps")
+	}
+
+	for i, svc := range svcs {
+		scopedLog := log.WithFields(logrus.Fields{
+			logfields.ServiceID: svc.FE.ID,
+			logfields.ServiceIP: svc.FE.L3n4Addr.String(),
+		})
+
+		if _, err := service.RestoreID(svc.FE.L3n4Addr, uint32(svc.FE.ID)); err != nil {
+			failed++
+			scopedLog.WithError(err).Warning("Unable to restore service ID")
+		}
+
+		for _, backend := range svc.BES {
+			s.backendRefCount.Add(backend.L3n4Addr.SHA256Sum())
+		}
+
+		s.svcByHash[svc.FE.SHA256Sum()] = svcs[i]
+		s.svcByID[svc.FE.ID] = svcs[i]
+		restored++
+	}
+
+	log.WithFields(logrus.Fields{
+		"restored": restored,
+		"failed":   failed,
+	}).Info("Restored services from maps")
+
+	return nil
+}
+
+func (s *Service) deleteServiceLocked(svc *lb.LBSVC) error {
+	obsoleteBackendIDs := s.deleteBackendsFromCacheLocked(svc)
+
+	if err := lbmap.DeleteService(svc.FE, len(svc.BES)); err != nil {
+		return err
+	}
+
+	delete(s.svcByHash, svc.Sha256)
+	delete(s.svcByID, svc.FE.ID)
+
+	ipv6 := svc.FE.L3n4Addr.IsIPv6()
+	for _, id := range obsoleteBackendIDs {
+		if err := lbmap.DeleteBackendByID(uint16(id), ipv6); err != nil {
+			return err
+		}
+	}
+	if err := service.DeleteID(uint32(svc.FE.ID)); err != nil {
+		return fmt.Errorf("Unable to release service ID %d: %s", svc.FE.ID, err)
+	}
+
+	deleteMetric.Inc()
+
+	return nil
+}
+
+func (s *Service) updateBackendsCacheLocked(svc *lb.LBSVC, backends []lb.LBBackEnd) ([]lb.LBBackEnd, []lb.BackendID, error) {
```

Comment:
Some unit tests for these functions would be :100:

Reply:
It's coming today (albeit in a separate PR)!